### ISSUE TITLE
Rule proposal: Prefer `.at()` method for negative index access

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#983]
https://github.com/tc39/proposal-relative-indexing-method

~~We cannot do this one yet because ESLint only supports stage 4 features.~~

What should the rule name be? `prefer-at-method-for-negative-index-access` is the most correct, but also a mouthful.

## Fail

```js
array[array.length - 1];
array.slice(-1)[0];
```

## Pass

```js
array.at(-1);
```

**Comment by fregante** [comment]
> We cannot do this one yet because ESLint only supports stage 4 features.

Being a non-syntax change, why would that be a problem?

**Comment by sindresorhus** [comment]
It would indeed not be a problem.

**Comment by fisker** [comment]
Accepted.

**Comment by fisker** [comment]
How about just `prefer-at`?

**Comment by fregante** [comment]
`prefer-array-at`?

**Comment by fisker** [comment]
It doesn't has  to be array https://github.com/tc39/proposal-relative-indexing-method#convertable-interfaces

**Comment by sindresorhus** [comment]
@fregante It's not just for Array.

**Comment by sindresorhus** [comment]
> How about just `prefer-at`?

👍

**Comment by fisker** [comment]
I think this rule  should also enforce `array.at(-2)` over `array.at(array.length - 2)`

**Comment by sindresorhus** [comment]
Alternatively, `prefer-at-method` or `prefer-at-index-access`.

**Comment by fisker** [comment]
> I think this rule should also enforce `array.at(-2)` over `array.at(array.length - 2)`

I'm going to add this to [`prefer-negative-index`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-negative-index.md) rule.

**Comment by fisker** [comment]
Should we check `string.charAt(string.length - 1)`? 

Even if we do, this won't be fixable, because `''.charAt(0)` returns empty string, and `''.at(0)` returns `undefined`

**Comment by sindresorhus** [comment]
> Should we check string.charAt(string.length - 1)?

Yes, it's better to always use `.at()` for consistency.

> Even if we do, this won't be fixable, because ''.charAt(0) returns empty string, and ''.at(0) returns undefined

Could we fix it to `'foo'..at(0) ?? ''`?

**Comment by fisker** [comment]
As I understand `.charAt` not very common used, especially for  negative index, I think suggestion is good enough.

**Comment by sindresorhus** [comment]
I think we should add an option to also include non-negative access too. There are many benefits over `array[1]`:

1. Supports negative index.
2. Consistency, instead of using both `array[1]` and `.at(-1)`. You use one way for index access.
3. `.at()` only accepts a number, `array[...]` happily stringifies the input, which could hide bugs.

**Comment by fisker** [comment]
Make sense, what option name should it be? `checkAllIndexAccess`?

**Comment by sindresorhus** [comment]
> checkAllIndexAccess?

Yes. When enabled, it should also check `.charAt(1)` and `.charAt()` (non-negative)`.

**Comment by fisker** [comment]
Added in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1331/commits/e58d47e90a117229fc82ff9a9a660b2eaffa6697,

When `checkAllIndexAccess` set to `true`:

- For computed access, only if we are sure it's a positive integer, use `.at()`
- For `String.charAt()`, always report with suggestion to use `.at()`
